### PR TITLE
Add support for bounded Redisson map cache (max size configuration). 

### PR DESCRIPTION
**Pull request by johnou**
Currently RedissonMapCache creates an unbounded set in Redis which may cause problems in certain cache implementations.

Thanks for the heavy lifting @zuanoc !

### CODE REVIEW

File: redisson/src/test/java/org/redisson/RedissonMapCacheTest.java [hunk 1]
```diff
@@ -213,6 +213,31 @@ public void testWriterFastPutTTL() {
         expected.put("3", "33");
         assertThat(store).isEqualTo(expected);
     }
+
+    @Test
+    public void testFastPutMaxSize() {
+        final int maxSize = 2;
+        Map<String, String> store = new LinkedHashMap<String, String>() {
+            @Override
+            protected boolean removeEldestEntry(Entry<String, String> eldest) {
```
Comment (johnou) [inline redisson/src/test/java/org/redisson/RedissonMapCacheTest.java]: required as store represents an onheap cache, not the cache in redis.

File: redisson/src/main/java/org/redisson/RedissonMapCache.java [hunk 2]
```diff
@@ -82,134 +81,161 @@
...
+                        "                    redis.call('zadd', KEYS[3], t + currentTime, key);" +
+                        "                else" +
+                        "                    map[i] = false;" +
+                        "                end;" +
+                        "            end;" +
+                        "        end;" +
+                        "    end;" +
+                        "end;" +
+                        "return map;",
                 Arrays.<Object>asList(getName(), getTimeoutSetName(), getIdleSetName()), args.toArray());
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/RedissonMapCache.java]: @johnou getLastAccessTimeSetName() is missed

File: redisson/src/main/java/org/redisson/RedissonMapCache.java [hunk 3]
```diff
@@ -334,72 +385,81 @@ protected boolean condition(Future<V> future) {
...
+                                "end;"
+                            + "redis.call('hdel', KEYS[1], ARGV[2]); "
+                            + "local msg = struct.pack('Lc0Lc0', string.len(ARGV[2]), ARGV[2], string.len(val), val); "
+                            + "redis.call('publish', KEYS[4], msg); "
+                            + "return 1; "
+                        + "else "
+                            + "return 0; "
+                        + "end",
+                Arrays.<Object>asList(getName(key), getTimeoutSetNameByKey(key), getIdleSetNameByKey(key), getRemovedChannelNameByKey(key),
+                        getLastAccessTimeSetName()),
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/RedissonMapCache.java]: @johnou getLastAccessTimeSetNameByKey should be used instead

File: redisson/src/main/java/org/redisson/RedissonMapCache.java [hunk 4]
```diff
@@ -261,68 +287,93 @@ public V putIfAbsent(K key, V value, long ttl, TimeUnit ttlUnit, long maxIdleTim
...
+                            + "redis.call('publish', KEYS[4], msg); "
+
+                            + "return nil;"
+                        + "else "
+                            + "local t, val = struct.unpack('dLc0', value); "
+                            + "redis.call('zadd', KEYS[3], t + ARGV[1], ARGV[5]); "
+                            + "return val;"
+                        + "end; ",
+                Arrays.<Object>asList(getName(key), getTimeoutSetNameByKey(key), getIdleSetNameByKey(key), getCreatedChannelNameByKey(key),
+                        getLastAccessTimeSetName(), getRemovedChannelName()),
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/RedissonMapCache.java]: @johnou getLastAccessTimeSetNameByKey, getRemovedChannelNameByKey should be used instead

File: redisson/src/main/java/org/redisson/RedissonMapCache.java [hunk 5]
```diff
@@ -1138,62 +1399,87 @@ public boolean fastPutIfAbsent(K key, V value, long ttl, TimeUnit ttlUnit, long
...
-                + "local msg = struct.pack('Lc0Lc0', string.len(ARGV[5]), ARGV[5], string.len(ARGV[6]), ARGV[6]); "
-                + "redis.call('publish', KEYS[4], msg); "
-				
-				+ "return 1; "
-			+ "else "
-				+ "return 0; "
-			+ "end; ",
-			Arrays.<Object>asList(getName(key), getTimeoutSetNameByKey(key), getIdleSetNameByKey(key), getCreatedChannelNameByKey(key)), 
-			System.currentTimeMillis(), ttlTimeout, maxIdleTimeout, maxIdleDelta, encodeMapKey(key), encodeMapValue(value));
+                "local insertable = false; "
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/RedissonMapCache.java]: @johnou please format this lua-script properly

File: redisson/src/main/java/org/redisson/RedissonMapCache.java [hunk 6]
```diff
@@ -1207,70 +1493,78 @@ protected boolean condition(Future<Boolean> future) {
...
-                "local expireDateScore = redis.call('zscore', KEYS[2], ARGV[2]); "
-                + "if expireDateScore ~= false then "
-                    + "expireDate = tonumber(expireDateScore) "
-                + "end; "
-
-                + "local t, val = struct.unpack('dLc0', v); "
-                + "if t ~= 0 then "
-                    + "local expireIdle = redis.call('zscore', KEYS[3], ARGV[2]); "
-                    + "if expireIdle ~= false then "
+                "local v = redis.call('hget', KEYS[1], ARGV[2]); "
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/RedissonMapCache.java]: @johnou please format this lua-script properly

File: redisson/src/main/java/org/redisson/RedissonMapCache.java [hunk 7]
```diff
@@ -1207,70 +1493,78 @@ protected boolean condition(Future<Boolean> future) {
...
-                
-                    + "local value = struct.pack('dLc0', t, string.len(ARGV[3]), ARGV[3]); "
-                    + "redis.call('hset', KEYS[1], ARGV[2], value); "
-                    
-                    + "local msg = struct.pack('Lc0Lc0Lc0', string.len(ARGV[2]), ARGV[2], string.len(ARGV[3]), ARGV[3], string.len(val), val); "
-                    + "redis.call('publish', KEYS[4], msg); "
-                    
-                    + "return val; ",
-                Arrays.<Object>asList(getName(key), getTimeoutSetNameByKey(key), getIdleSetNameByKey(key), getUpdatedChannelNameByKey(key)), 
+                "local value = redis.call('hget', KEYS[1], ARGV[2]); "
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/RedissonMapCache.java]: @johnou please format this lua-script properly

**Comment by zuanoc** [comment]
@mrniko PTAL

**Comment by mrniko** [comment]
@zuanoc seems you fixed only one case

**Comment by mrniko** [comment]
@johnou please merge it with "master" branch

**Comment by johnou** [comment]
@mrniko https://github.com/redisson/redisson/pull/1027

**Pull request merged**